File: DS using c++/linked-lists/singlyLL.cpp
#include <algorithm>
#include <cstdio>
#include <iomanip>
#include <iostream>

using namespace std;

const char line = '\n';

class Node {
   public:
    string name;
    string rollnumber;
    Node *next;
    Node(string name = "", string rollnumber = "",
         Node *next = NULL) {
        this->name = name;
        this->rollnumber = rollnumber;
        this->next = next;
    }
};

class Linked_List {
   public:
    Node *head /*=new Node()*/;
    Node *first /*=new Node()*/;
    Node *last /*=new Node()*/;
    void intialization_of_heads() {
        string name, rollnumber;
        cout << "Enter the name of the readers club head : ";
        getline(cin, name);
        transform(name.begin(), name.end(), name.begin(), ::toupper);
        cout << "Enter the roll number of the readers club head : ";
        getline(cin, rollnumber);
        transform(rollnumber.begin(), rollnumber.end(),
                  rollnumber.begin(), ::toupper);
        first = new Node(name, rollnumber);
        cout << "Enter the name of the in-charge of readers club : ";
        getline(cin, name);
        transform(name.begin(), name.end(), name.begin(), ::toupper);
        cout << "Enter the roll number of the in-charge of readers "
                "club : ";
        getline(cin, rollnumber);
        transform(rollnumber.begin(), rollnumber.end(),
                  rollnumber.begin(), ::toupper);
        last = new Node(name, rollnumber);
        first->next = last;
        head = first;
    }

    Node *searchNode(string rollnumber) {
        Node *temp = NULL;
        Node *ptr = head;
        while (ptr != NULL) {
            if (ptr->rollnumber == rollnumber) {
                temp = ptr;
                break;
            }
            ptr = ptr->next;
        }
        return temp;
    }

    void add() {
        string rollnumber;
        getline(cin, rollnumber);
        cout << "Enter the roll number of person : ";
        getline(cin, rollnumber);
        transform(rollnumber.begin(), rollnumber.end(),
                  rollnumber.begin(), ::toupper);
        if (searchNode(rollnumber) != NULL) {
            cout
                << "\nStudent is already present in readers club\n\n";
        } else {
            string name;
            cout << "Enter the name of person : ";
            getline(cin, name);
            transform(name.begin(), name.end(), name.begin(),
                      ::toupper);
            if (head == NULL) {
                first->name = name;
                first->rollnumber = rollnumber;
                head = first;
                head->next = NULL;
                cout << "\nSuccessfully added the student to readers "
                        "club\n";
                return;
            }
            if (head->next == NULL) {
                last = new Node(name, rollnumber);
                head->next = last;
                cout << "\nSuccessfully added the student to readers "
                        "club\n";
                return;
            }
            Node *temp = head;
            while (temp->next != last) temp = temp->next;
            temp->next = new Node(name, rollnumber, last);
            cout << "\nSuccessfully added the students to readers "
                    "club\n";
        }
    }

    int total_members() {
        int n = 0;
        Node *temp = head;
        while (temp != NULL) {
            temp = temp->next;
            n++;
        }
        return n;
    }

    void sort() {
        string a;
        string b;
        Node *temp = head;
        while (temp->next != NULL) {
            Node *min = temp;
            Node *temp1 = temp->next;
            while (temp1 != NULL) {
                if (temp1->name < min->name) {
                    min = temp1;
                }
                temp1 = temp1->next;
            }
            swap(min->name, temp->name);
            swap(min->rollnumber, temp->rollnumber);
            temp = temp->next;
        }
    }

    void delet() {
        if (total_members() == 0) {
            cout << "\nThere are no members in readers club\n";
        } else {
            string rollnumber;
            getline(cin, rollnumber);
            cout << "Enter the roll number of student you want to "
                    "delete: ";
            getline(cin, rollnumber);
            transform(rollnumber.begin(), rollnumber.end(),
                      rollnumber.begin(), ::toupper);
            if (searchNode(rollnumber) == NULL) {
                cout << "\nThere is no one in the readers club with "
                        "the roll number "
                     << rollnumber << "\n";
                return;
            }
            sort();
            if (head->rollnumber == rollnumber) {
                head = head->next;
                /*last=NULL;*/
                delete last;
                cout << "\nThe student you entered has been removed "
                        "form readers club\n";
                return;
            }
            Node *temp = head;
            Node *prev = temp;
            while (temp->next != NULL &&
                   temp->rollnumber != rollnumber) {
                prev = temp;
                temp = temp->next;
                if (temp->rollnumber > rollnumber) break;
            }
            if (temp->rollnumber == rollnumber) {
                prev->next = temp->next;
                temp->next = NULL;
                delete temp->next;
            }
            cout << "\nThe student you entered has been removed form "
                    "readers club\n";
        }
    }

    void display_members() {
        Node *temp = head;
        printf("%-30s%s\n", "Name", "Roll Number");
        while (temp != NULL) {
            cout << temp->name << setw(30) << temp->rollnumber
                 << line;
            // printf("%-30s%s\n", temp->name, temp->rollnumber);
            temp = temp->next;
        }
        cout << line;
    }

    void display_reverse(Node *head) {
        if (head->next != NULL) {
            display_reverse(head->next);
        }
        cout << head->name << setw(30) << head->rollnumber << line;
    }

    void todo(int option) {
        switch (option) {
            case 1:
                add();
                break;
            case 2:
                delet();
                break;
            case 3:
                cout << "\nTotal number of members in readers club "
                        "are "
                     << total_members() << "\n";
                break;
            case 4:
                if (total_members() == 0)
                    cout
                        << "\nThere are no members in readers club\n";
                else {
                    cout << "Name " << setw(30) << " Rollnumber\n\n";
                    display_members();
                }
                break;
            case 5:
                if (total_members() == 0)
                    cout
                        << "\nThere are no members in readers club\n";
                else {
                    cout << "Name " << setw(30) << " Rollnumber\n\n";
                    display_reverse(head);
                }
                cout << "\n";
                break;
            case 6:
                if (total_members() == 0)
                    cout << "\nThere are no members in readers club "
                            "to sort\n";
                else {
                    sort();
                    cout << "\nSorted the list successfully\n";
                }
                break;
            case 7:
                cout << "\nProgram ended!!\n";
                break;
            default:
                string a;
                getline(cin, a);
                cout << "\nInvalid option, please try again\n";
        }
    }

    void operation() {
        int option;
        do {
            cout
                << ("\n1-> Add a member\n"
                    "2-> Delete a members\n"
                    "3-> Count of students in readers club\n"
                    "4-> Display All members in readers club\n"
                    "5-> Display all the members in reverse order\n"
                    "6-> sort the list\n"
                    "7-> To exit Enter 7 or any number >7\n");
            cout << ("\nEnter your option: ");
            cin >> option;
            todo(option);
        } while (option != 7);
    }
};
int main() {
    Linked_List l;
    l.intialization_of_heads();
    cout << line;
    l.operation();
    return 0;
}

File: DS using python/linked-lists/singlyLL.py
class Node:
    def __init__(self, name="", rollnumber="", next=None):
        self.name = name
        self.rollnumber = rollnumber
        self.next = next


class LinkedList:
    def __init__(self):
        self.head = None
        self.first = None
        self.last = None

    def searchNode(self, rollnumber):
        temp = None
        ptr = self.head
        while ptr:
            if ptr.rollnumber.upper() == rollnumber.upper():
                temp = ptr
                break
            ptr = ptr.next
        return temp

    def add(self):
        rollnumber = input("Enter the roll number of student : ")
        if self.searchNode(rollnumber):
            print("Student is already in the readers club\n")
            return
        name = input("Enter the name of student : ")
        if self.head == None:
            self.first.name, self.first.rollnumber = name, rollnumber
            self.head = self.first
            self.head.next = None
            print("\nSuccessfully added the student to readers club\n")
            return

        if self.head.next is None:
            self.last = Node(name, rollnumber)
            self.head.next = self.last
            print("\nSuccessfully added the student to readers club\n")
            return
        temp = self.head
        while temp.next != self.last:
            temp = temp.next
        temp.next = Node(name, rollnumber, self.last)
        print("\nSuccessfully added the student to readers club\n")

    def delete(self):
        if self.totalMembers() == 0:
            print("\nThere are no members in readers club\n")
            return
        rollnumber = input("Enter the roll number of student you want to delete : ")
        if self.searchNode(rollnumber) is None:
            print("\nMember not found in readers club\n")
            return
        if self.head.rollnumber == rollnumber:
            self.head = self.head.next
            self.last = None
            print("\nThe student you entered has been removed form readers club\n")
            return
        temp = self.head
        prev = temp
        while temp != None and temp.rollnumber != rollnumber:
            prev = temp
            temp = temp.next
        if temp.rollnumber == rollnumber:
            prev.next = temp.next
            temp.next = None
        print("\nThe student you entered has been removed form readers club\n")

    def totalMembers(self):
        n = 0
        temp = self.head
        while temp:
            temp = temp.next
            n += 1
        return n

    def displayMembers(self):
        temp = self.head
        print("\n%-30s%s\n" % ("Name", "Roll Number"))
        while temp:
            print("%-30s%s\n" % (temp.name, temp.rollnumber))
            temp = temp.next
        print()

    def displayReverse(self, head):
        if head.next is not None:
            self.displayReverse(head.next)
        print("%-30s%s\n" % (head.name, head.rollnumber))

    def sort(self):
        temp = self.head
        while temp.next is not None:
            min = temp
            temp1 = temp.next
            while temp1 is not None:
                if temp1.name.upper() < min.name.upper():
                    min = temp1
                temp1 = temp1.next
            temp.name, temp.rollnumber, min.name, min.rollnumber = (
                min.name,
                min.rollnumber,
                temp.name,
                temp.rollnumber,
            )
            temp = temp.next

    def todo(self, option):
        if option in range(1, 7):
            if option == 1:
                self.add()
            elif option == 2:
                self.delete()
            elif option == 3:
                print(
                    "\nTotal number of members in readers club are %d\n\n"
                    % (self.totalMembers())
                )
            elif option == 4:
                if self.totalMembers() == 0:
                    print("\nThere are no members in readers club")
                else:
                    self.displayMembers()
            elif option == 5:
                if self.totalMembers() == 0:
                    print("\nThere are no members in readers club")
                else:
                    print("\n%s%30s\n\n" % ("Name", "Roll Number"))
                    self.displayReverse(self.head)
            elif option == 6:
                if self.totalMembers() == 0:
                    print("\nThere are no members in readers club")
                else:
                    self.sort()
                    print("\nSorted the list successfully\n")
            elif option == 7:
                print("Program ended!!\n")
        elif option not in range(1, 8):
            print("Invalid option, please try again\n")

    def operation(self):
        option = 0
        while option < 7:
            print(
                "1. Add a member\n"
                + "2. Delete a members\n"
                + "3. Count of students in readers club\n"
                + "4. Display All members in readers club\n"
                + "5. Display all the members in reverse order\n"
                + "6. sort the list\n"
                + "7. To exit Enter 7 or any number >7\n"
            )
            option = int(input("Enter your option: "))
            self.todo(option)

    def intialization(self):
        self.first = Node(
            input("Enter the name of Reader's club head : "),
            input("Enter the roll number of Reader's club head: "),
        )
        self.last = Node(
            input("Enter the name of Reader's club in-charge : "),
            input("Enter the roll number of Reader's club in-charge : "),
        )
        self.first.next = self.last
        self.head = self.first


if __name__ == "__main__":
    l = LinkedList()
    l.intialization()
    print()
    l.operation()
